Docs/connectivity (#644)
dbbs-lab/bsb: docs/connectivity/connectivity-strategy.rst

          granule_close_enough = dist < self.radius
          
          #We find the indices of the 40 closest granule cells
          to_connect_ids = np.argsort(granule_close_enough)[0:self.divergence]

          #Since we are interested in connecting point-like cells, we do not need to store
          #info about the precise position on the dendrites or axons; 
          #It is enough to store which presynaptic cell is connetcted to
          #certain postsynaptic cells, namely the first entry of both pre_set and post_set.
          
          #The index of the presynaptic cell in the golgi_pos array is i
          pre_set[ptr:ptr+self.divergence,0] = i
          #We store in post_set the indices of the postsynaptic cells we selected before.
          post_set[ptr:ptr+self.divergence,0] = to_connect_ids
          ptr += to_be_connected

        #Now we connect the cells according to the information stored in src_locs and dest_locs
        #calling the connect_cells method.
        connect_cells(pre_set, post_set, src_locs, dest_locs)

Use case 2 : Connection between a detailed cell and a point-like cell. 
======================================================================

> If we have a detailed morphology of the pre or post synaptic cells we can specify where to form the connection. Suppose we want to connect Golgi cells to glomeruli specifying the position of the connection on the Golgi cell axon. In this example we form a connection on the closest point to a glomerulus.

Reviewer: ```suggestion
If we have a detailed morphology of the pre- or postsynaptic cells we can specify where to form the connection. Suppose we want to connect Golgi cells to glomeruli specifying the position of the connection on the Golgi cell axon. In this example we form a connection on the closest point to a glomerulus.
```

First, we need to specify the neurites where we want to form a synapse in the configuration file, using the "morphology_labels" tag. 

.. code-block:: json
  
  "golgi_to_granule": {
        "strategy": "cerebellum.connectome.golgi_granule.ConnectomeGolgiGranule",
        "radius": 100,
        "convergence": 40,
        "presynaptic": {
          "cell_types": ["glomerulus"]
        },
        "postsynaptic": {
          "cell_types": ["golgi_cell"],
          "morphology_labels" : ["basal_dendrites"]
        }
      }

> The :meth:`~.bsb.connectivity.strategy.ConnectionStrategy.get_region_of_interest` is analogous to the previous example, so we focus only on the :meth:`~.bsb.connectivity.strategy.ConnectionStrategy.connect` method.

Reviewer: ```suggestion
The :meth:`~bsb.connectivity.strategy.ConnectionStrategy.get_region_of_interest` is analogous to the previous example, so we focus only on the :meth:`~bsb.connectivity.strategy.ConnectionStrategy.connect` method.
```

        glomeruli_pos = post_ps.load_positions()
        n_glomeruli = len(glomeruli_pos)
        n_golgi = len(golgi_pos)
        n_conn = n_glomeruli * n_golgi
        # We define two arrays of length n_conn to store the connections to be made,
        # even if we will not use all the entries of arrays, for the sake of speed.
        # We keep track of how many entries, namely how many connection, by means of the variable ptr.
        pre_locs = np.full((n_conn, 3), -1, dtype=int)
        post_locs = np.full((n_conn, 3), -1, dtype=int)
        ptr = 0
      
        #Cache morphologies and generate the morphologies iterator.
        morpho_set = post_ps.load_morphologies()
        golgi_morphos = morpho_set.iter_morphologies(cache=True, hard_cache=True)
        
        #Loop through all the Golgi cells
        for i, golgi, morpho in zip(itertools.count(), golgi_pos, golgi_morphos):
            
            #We compute the distance between the current Golgi cell and all the glomeruli,
            #then select the good ones.
            dist = np.sqrt(
                np.power(golgi[0] - glomeruli_pos[:, 0], 2)
                + np.power(golgi[1] - glomeruli_pos[:, 1], 2)

>                 + np.power(golgi[2] - glomeruli_pos[:, 2], 2)
>             )
>             
>             to_connect_bool = dist < self.radius
>             to_connect_idx = np.nonzero(to_connect_bool)[0]
>             connected_gloms = len(to_connect_idx)
>             
>             #We assign the indices of the Golgi cell and the granule cells to connect 
>             pre_locs[ptr : (ptr + connected_gloms), 0] = to_connect_idx
>             post_locs[ptr : (ptr + connected_gloms), 0] = i
>            
>             #Get the branches corresponding to basal dendrites.
>             #morpho contains only the branches tagged as specified 
>             #in the configuration file.
>             basal_dendrides_branches = morpho.get_branches()
>             
>             #Get the starting branch id of the denridic branches
>             first_dendride_id = morpho.branches.index(basal_dendrides_branches[0])
>             
>             #Find terminal points on branches
>             terminal_ids = np.full(len(basal_dendrides_branches), 0, dtype=int)
>             for i,b in enumerate(basal_dendrides_branches):
>                 if b.is_terminal:
>                     terminal_ids[i] = 1
>             terminal_branches_ids = np.nonzero(terminal_ids)[0]
> 
>             #Keep only terminal branches
>             basal_dendrides_branches = np.take(basal_dendrides_branches, terminal_branches_ids, axis=0)
>             terminal_branches_ids = terminal_branches_ids + first_dendride_id
> 
>             #Find the point-on-branch ids of the tips
>             tips_coordinates = np.full((len(basal_dendrides_branches),3), 0, dtype=float)
>             for i,branch in enumerate(basal_dendrides_branches):
>                 tips_coordinates[i] = branch.points[-1]
> 
>             #Choose randomly the branch where the synapse is made
>             #favouring the branches closer to the glomerulus.
>             rolls = exp_dist.rvs(size=len(basal_dendrides_branches))
>             
>             # Compute the distance between terminal points of basal dendrites 
>             # and the soma of the avaiable glomeruli
>             for id_g,glom_p in enumerate(glomeruli_pos):
>                 pts_dist = np.sqrt(np.power(tips_coordinates[:,0] + golgi[0] - glom_p[0], 2)
>                         + np.power(tips_coordinates[:,1] + golgi[1] - glom_p[1], 2)
>                         + np.power(tips_coordinates[:,2] + golgi[2] - glom_p[2], 2)
>                     )
> 
>                 sorted_pts_ids = np.argsort(pts_dist)
>                 # Pick the point in which we form a synapse according to a exponential distribution mapped
>                 # through the distance indices: high chance to pick closeby points.
>                 pt_idx = sorted_pts_ids[int(len(basal_dendrides_branches)*rolls[np.random.randint(0,len(rolls))])]
> 
>                 #The id of the branch is the id of the terminal_branches plus the id of the first dendritic branch
>                 post_locs[ptr+id_g,1] = terminal_branches_ids[pt_idx]
>                 #We connect the tip of the branch
>                 post_locs[ptr+id_g,2] = len(basal_dendrides_branches[pt_idx].points)-1
>             ptr += connected_gloms
> 
>         #Now we connect the cells
>         self.connect_cells(pre_ps, post_ps, pre_locs[:ptr], post_locs[:ptr])

Reviewer: ```suggestion
    def connect(self, pre, post):
      for pre_ct, pre_ps in pre.placement.items():
          for post_ct, post_ps in post.placement.items():
              self._connect_type(pre_ct, pre_ps, post_ct, post_ps)
  
      def _connect_type(self, pre_ct, pre_ps, post_ct, post_ps):
        # We store the positions of the pre and post synaptic cells.
        golgi_pos = pre_ps.load_positions()
        glomeruli_pos = post_ps.load_positions()
        n_glomeruli = len(glomeruli_pos)
        n_golgi = len(golgi_pos)
        max_conn = n_glomeruli * n_golgi
        # We define two arrays of length `max_conn ` to store the connections,
        # even if we will not use all the entries of arrays, for the sake of speed.
        pre_locs = np.full((max_conn , 3), -1, dtype=int)
        post_locs = np.full((max_conn , 3), -1, dtype=int)
        # `ptr` keeps track of how many connections we've made so far.
        ptr = 0
      
        # Cache morphologies and generate the morphologies iterator.
        morpho_set = post_ps.load_morphologies()
        golgi_morphos = morpho_set.iter_morphologies(cache=True, hard_cache=True)
        
        # Loop through all the Golgi cells
        for i, golgi, morpho in zip(itertools.count(), golgi_pos, golgi_morphos):
            
            # We compute the distance between the current Golgi cell and all the glomeruli,
            # then select the good ones.
            dist = np.sqrt(
                np.power(golgi[0] - glomeruli_pos[:, 0], 2)
                + np.power(golgi[1] - glomeruli_pos[:, 1], 2)
                + np.power(golgi[2] - glomeruli_pos[:, 2], 2)
            )
            
            to_connect_bool = dist < self.radius
            to_connect_idx = np.nonzero(to_connect_bool)[0]
            connected_gloms = len(to_connect_idx)
            
            # We assign the indices of the Golgi cell and the granule cells to connect 
            pre_locs[ptr : (ptr + connected_gloms), 0] = to_connect_idx
            post_locs[ptr : (ptr + connected_gloms), 0] = i
           
            # Get the branches corresponding to basal dendrites.
            # `morpho` contains only the branches tagged as specified 
            # in the configuration file.
            basal_dendrides_branches = morpho.get_branches()
            
            # Get the starting branch id of the denridic branches
            first_dendride_id = morpho.branches.index(basal_dendrides_branches[0])
            
            # Find terminal points on branches
            terminal_ids = np.full(len(basal_dendrides_branches), 0, dtype=int)
            for i,b in enumerate(basal_dendrides_branches):
                if b.is_terminal:
                    terminal_ids[i] = 1
            terminal_branches_ids = np.nonzero(terminal_ids)[0]

            # Keep only terminal branches
            basal_dendrides_branches = np.take(basal_dendrides_branches, terminal_branches_ids, axis=0)
            terminal_branches_ids = terminal_branches_ids + first_dendride_id

            # Find the point-on-branch ids of the tips
            tips_coordinates = np.full((len(basal_dendrides_branches),3), 0, dtype=float)
            for i,branch in enumerate(basal_dendrides_branches):
                tips_coordinates[i] = branch.points[-1]

            # Choose randomly the branch where the synapse is made
            # favouring the branches closer to the glomerulus.
            rolls = exp_dist.rvs(size=len(basal_dendrides_branches))
            
            # Compute the distance between terminal points of basal dendrites 
            # and the soma of the avaiable glomeruli
            for id_g,glom_p in enumerate(glomeruli_pos):
                pts_dist = np.sqrt(np.power(tips_coordinates[:,0] + golgi[0] - glom_p[0], 2)
                        + np.power(tips_coordinates[:,1] + golgi[1] - glom_p[1], 2)
                        + np.power(tips_coordinates[:,2] + golgi[2] - glom_p[2], 2)
                    )

                sorted_pts_ids = np.argsort(pts_dist)
                # Pick the point in which we form a synapse according to a exponential distribution mapped
                # through the distance indices: high chance to pick closeby points.
                pt_idx = sorted_pts_ids[int(len(basal_dendrides_branches)*rolls[np.random.randint(0,len(rolls))])]

                # The id of the branch is the id of the terminal_branches plus the id of the first dendritic branch
                post_locs[ptr+id_g,1] = terminal_branches_ids[pt_idx]
                # We connect the tip of the branch
                post_locs[ptr+id_g,2] = len(basal_dendrides_branches[pt_idx].points)-1
            ptr += connected_gloms

        # Now we connect the cells
        self.connect_cells(pre_ps, post_ps, pre_locs[:ptr], post_locs[:ptr])
```
